src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -984,4 +1082,205 @@ mod tests {
+            hash: Bytes::from_str(
+                "0x035c0e674c3bf3384a74b766908ab41c1968e989360aa26bea1dd64b1626f5f0",
+            )
+            .unwrap(),
+            timestamp: 1748397011,
+            ..Default::default()
+        };
+
+        let usv4_state = UniswapV4State::new(
+            u128::MAX / 2,
+            U256::from_str("79228162514264337593543950336").unwrap(),
+            UniswapV4Fees { zero_for_one: 100, one_for_zero: 100, lp_fee: 100 }, // Low fees
+            0,
+            60,
+            vec![
+                TickInfo::new(-600, (u128::MAX / 4) as i128),
+                TickInfo::new(600, -((u128::MAX / 4) as i128)),
+            ],
+            block,
+        );
+
+        let token_in = usdc();
+        let token_out = weth();
+
+        let max_amount = usv4_state
+            .find_max_amount(&token_in, &token_out)
+            .unwrap();
+        assert!(max_amount > BigUint::from(1u64));
+        assert!(usv4_state
+            .get_amount_out(max_amount.clone(), &token_in, &token_out)
+            .is_ok());
+        assert!(usv4_state
+            .get_amount_out(max_amount.clone() + BigUint::from(1u64), &token_in, &token_out)
+            .is_err());
+    }
+
+    #[test]
+    fn test_find_max_amount_less_than_starting_point() {
+        // We start our search at 1e18, so test that we can properly find a max amount less than
+        // that
+
+        let block = BlockHeader {
+            number: 22578103,
+            hash: Bytes::from_str(
+                "0x035c0e674c3bf3384a74b766908ab41c1968e989360aa26bea1dd64b1626f5f0",
+            )
+            .unwrap(),
+            timestamp: 1748397011,
+            ..Default::default()
+        };
+
+        let usv4_state = UniswapV4State::new(
+            1000000000000000000u128, // 10e17
+            U256::from_str("79228162514264337593543950336").unwrap(),
+            UniswapV4Fees { zero_for_one: 100, one_for_zero: 100, lp_fee: 3000 },
+            0,
+            60,
+            vec![
+                TickInfo::new(-180, 5000000000000000000i128),
+                TickInfo::new(180, -5000000000000000000i128),
+            ],
+            block,
+        );
+
+        let token_in = usdc();
+        let token_out = weth();
+
+        let max_amount = usv4_state
+            .find_max_amount(&token_in, &token_out)
+            .unwrap();
+
+        let success = usv4_state
+            .get_amount_out(max_amount.clone(), &token_in, &token_out)
+            .is_ok();
+        assert!(success, "Should be able to swap the exact max amount");
+
+        let one_more = &max_amount + BigUint::from(1u64);
+        let should_fail = usv4_state
+            .get_amount_out(one_more, &token_in, &token_out)
+            .is_err();
+        assert!(should_fail, "Swapping max_amount + 1 should fail");
+    }
+
+    #[test]
+    fn test_find_max_amount_boundary_conditions() {
+        // Test edge cases and boundary conditions
```

Comment:
what exactly are the edge cases and boundary conditions here?

Reply:
I've done a test refactor to parametrize this actually, it's a lot cleaner now. Have a look at my latest commit!